### PR TITLE
Feature/template repo contents

### PR DESCRIPTION
**Pull request by darrylmorton-digicatapult**
Template repo contents
- [x] service template repo 
- [x] openapi
- [x] helm
- [x] auth token verification
- [x] github actions 
- [x] empty github-pages branch
- [x] VITALam team created and granted repo access

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        uses: azure/setup-helm@v1
+        with:
+          version: v3.4.0
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.7
+      - name: Set up chart-testing
+        uses: helm/chart-testing-action@v2.0.1
+      - name: Run chart-testing (lint)
+        run: ct lint --validate-maintainers false --charts helm/${{ needs.preconditions.outputs.repo_name }}/ --chart-repos bitnami=https://charts.bitnami.com/bitnami,k8s-at-home=https://k8s-at-home.com/charts
```
Comment (dblane-digicatapult) [inline .github/workflows/release.yml]: At this stage there are no additional chart repos required.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+          tags: |
+            ${{ needs.preconditions.outputs.repo_name }}:latest
+          labels: |
+            org.opencontainers.image.title=${{ needs.preconditions.outputs.repo_name }}
+            org.opencontainers.image.description=${{ github.event.repository.description }}
+            org.opencontainers.image.source=${{ github.event.repository.html_url }}
+            org.opencontainers.image.url=${{ github.event.repository.html_url }}
+            org.opencontainers.image.revision=${{ github.sha }}
+      - name: Run chart-testing (install)
+        run: ct install  --charts helm/${{ needs.preconditions.outputs.repo_name }}/ --chart-repos bitnami=https://charts.bitnami.com/bitnami,k8s-at-home=https://k8s-at-home.com/charts
```
Comment (dblane-digicatapult) [inline .github/workflows/release.yml]: Again no additional chart repos required.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        uses: azure/setup-helm@v1
+        with:
+          version: v3.4.0
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.7
+      - name: Set up chart-testing
+        uses: helm/chart-testing-action@v2.0.1
+      - name: Run chart-testing (lint)
+        run: ct lint --validate-maintainers false --charts helm/${{ needs.repo_ids.outputs.repo_name }}/ --chart-repos bitnami=https://charts.bitnami.com/bitnami,k8s-at-home=https://k8s-at-home.com/charts
```
Comment (dblane-digicatapult) [inline .github/workflows/tests.yml]: remove chart-repos

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+          tags: |
+            ${{ needs.repo_ids.outputs.repo_name }}:latest
+          labels: |
+            org.opencontainers.image.title=${{ needs.repo_ids.outputs.repo_name }}
+            org.opencontainers.image.description=${{ github.event.repository.description }}
+            org.opencontainers.image.source=${{ github.event.repository.html_url }}
+            org.opencontainers.image.url=${{ github.event.repository.html_url }}
+            org.opencontainers.image.revision=${{ github.sha }}
+      - name: Run chart-testing (install)
+        run: ct install --charts helm/${{ needs.repo_ids.outputs.repo_name }}/ --chart-repos bitnami=https://charts.bitnami.com/bitnami,k8s-at-home=https://k8s-at-home.com/charts
```
Comment (dblane-digicatapult) [inline .github/workflows/tests.yml]: remove chart-repos

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+# from the existing environment variable, otherwise the environment variable
+# shadows the argument.
+ARG LOGLEVEL
+ENV NPM_CONFIG_LOGLEVEL ${LOGLEVEL}
+
+WORKDIR /vitalam-service-template
+
+# Install base dependencies
+COPY . .
+RUN npm install --production
```
Comment (dblane-digicatapult) [inline Dockerfile]: Update to new version of `npm` so we use new package-lock format.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+## Environment Variables
+
+`vitalam-service-template` is configured primarily using environment variables as follows:
+
+|           variable            | required | default | description                                                                                                          |
+| :---------------------------- | :------: | :-----: | :------------------------------------------------------------------------------------------------------------------- |
+| SERVICE_TYPE                  |    N     | `info`  | Logging level. Valid values are [`trace`, `debug`, `info`, `warn`, `error`, `fatal`]                                 |
+| PORT                          |    N     | `3001`  | The port for the API to listen on                                                                                    |
+| API_HOST                      |    Y     |    -    | The hostname of the `vitalam-node` the API should connect to                                                         |
+| API_PORT                      |    N     | `9944`  | The port of the `vitalam-node` the API should connect to                                                             |
```
Comment (dblane-digicatapult) [inline README.md]: Rename to API_RPC_PORT as the vitalam-node has a websocket API and a generic RPC API port

File: app/env.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  dotenv.config({ path: '.env' })
+}
+
+const vars = envalid.cleanEnv(
+  process.env,
+  {
+    SERVICE_TYPE: envalid.str({ default: 'vitalam-service-template'.toUpperCase().replace(/-/g, '_') }),
+    PORT: envalid.port({ default: 80, devDefault: 3000 }),
+    API_HOST: envalid.host({ devDefault: 'localhost' }),
+    API_PORT: envalid.port({ default: 9944 }),
```
Comment (dblane-digicatapult) [inline app/env.js]: Rename as API_RPC_PORT

File: app/util/appUtil.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
+const { ApiPromise, WsProvider, Keyring } = require('@polkadot/api')
+const jwt = require('jsonwebtoken')
+const {
+  API_HOST,
+  API_PORT,
```
Comment (dblane-digicatapult) [inline app/util/appUtil.js]: API_RPC_PORT

File: docker-compose.yaml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    image: ghcr.io/digicatapult/vitalam-node:v2.3.0
+    ports:
+      - 30333:30333
+      - 9944:9944
+      - 9933:9933
+    environment:
+      - IDENTITY=dev
+      - WS=true
+      - RPC=true
+      - CORS=all
```
Comment (dblane-digicatapult) [inline docker-compose.yaml]: I believe we now need to change the `CMD` that is run on the vitalam-node container as we no longer use ENV variables and a `docker-run.sh` file to insert these as arguments.

File: test/test.env [hunk 10]
```diff
@@ -0,0 +1,10 @@
+LOG_LEVEL=fatal
+
+PORT=3000
+API_HOST=localhost
+API_PORT=9944
```
Comment (dblane-digicatapult) [inline test/test.env]: API_RPC_PORT

**Review by dblane-digicatapult — CHANGES_REQUESTED** [review]

File: docker-compose.yaml (hunk 9 above)
Comment (jonmattgray) [reply docker-compose.yaml]: made a bug ticket for this earlier https://digicatapult.atlassian.net/browse/IN-65

File: app/env.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    API_VERSION: envalid.str({ default: version }),
+    API_MAJOR_VERSION: envalid.str({ default: 'v1' }),
+    LOG_LEVEL: envalid.str({ default: 'info', devDefault: 'debug' }),
+    USER_URI: envalid.str({ devDefault: '//Alice' }),
+    AUTH_JWKS_URI: envalid.url(),
+    AUTH_AUDIENCE: envalid.str(),
+    AUTH_ISSUER: envalid.url(),
+    AUTH_TOKEN_URL: envalid.url(),
```
Comment (jonmattgray) [inline app/env.js]: the envs from here down can go?

Comment (darrylmorton-digicatapult) [reply app/env.js]: removed the authentication config and deps

File: docker-compose.yaml (hunk 9 above)
Comment (mattdean-digicatapult) [reply docker-compose.yaml]: I don't see why the service template even needs the `vitalam-node`?

File: app/util/appUtil.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
+const { ApiPromise, WsProvider } = require('@polkadot/api')
```
Comment (jonmattgray) [inline app/util/appUtil.js]: if we're not including the node in the template, this whole file can go + envs for node config + polkadot/api package

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+coverage
+.DS_Store
+.nyc_output
+.idea
+.env
+.vscode
+
+github_token
+helm/**/*.tgz
+jsHelper.js
```
Comment (danielcooperxyz) [inline .gitignore]: Is this a common file to ignore in projects?

Comment (mattdean-digicatapult) [reply .gitignore]: No. I don't know why this is ignored when it's not referenced anywhere

Comment (darrylmorton-digicatapult) [reply .gitignore]: > Is this a common file to ignore in projects?

good question, it's not by convention and something I started adding for convenience which is still yet to be discussed.
Happy to remove this if this is not something that is useful to everyone. Thoughts team?

File: helm/vitalam-service-template/templates/deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+              valueFrom:
+                configMapKeyRef:
+                  name: {{ include "vitalam-service-template.fullname" . }}-config
+                  key: port
+            - name: API_HOST
+              valueFrom:
+                configMapKeyRef:
+                  name: {{ include "vitalam-service-template.fullname" . }}-config
+                  key: apiHost
+            - name: USER_URI
```
Comment (jonmattgray) [inline helm/vitalam-service-template/templates/deployment.yaml]: old env for node, can be removed, in configmap and values too

File: test/test.env [hunk 15]
```diff
@@ -0,0 +1,6 @@
+LOG_LEVEL=fatal
+
+PORT=3000
+API_HOST=localhost
+API_RPC_PORT=9944
```
Comment (jonmattgray) [inline test/test.env]: these last two are node envs

File: README.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+## Environment Variables
+
+`vitalam-service-template` is configured primarily using environment variables as follows:
+
+| variable                      | required | default | description                                                                                     |
+| :---------------------------- | :------: | :-----: | :---------------------------------------------------------------------------------------------- |
+| SERVICE_TYPE                  |    N     | `info`  | Logging level. Valid values are [`trace`, `debug`, `info`, `warn`, `error`, `fatal`]            |
+| PORT                          |    N     | `3001`  | The port for the API to listen on                                                               |
+| API_HOST                      |    Y     |    -    | The hostname of the `vitalam-node` the API should connect to                                    |
+| API_RPC_PORT                  |    N     | `9944`  | The port of the `vitalam-node` the API should connect to                                        |
```
Comment (jonmattgray) [inline README.md]: some node envs to remove

File: app/env.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+} else {
+  dotenv.config({ path: '.env' })
+}
+
+const vars = envalid.cleanEnv(
+  process.env,
+  {
+    SERVICE_TYPE: envalid.str({ default: 'vitalam-service-template'.toUpperCase().replace(/-/g, '_') }),
+    PORT: envalid.port({ default: 80, devDefault: 3000 }),
+    API_HOST: envalid.host({ devDefault: 'localhost' }),
```
Comment (mattdean-digicatapult) [inline app/env.js]: `API_HOST` is I believe unused

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 13 above)
Comment (danielcooperxyz) [reply .gitignore]: Not sure it's a good idea to have files mentioned here and in the linter config, that don't exist and may be created or have code added in the future?

Comment (mattdean-digicatapult) [reply .gitignore]: Absoltely these lines should be removed

File: .eslintignore [hunk 18]
```diff
@@ -0,0 +1,5 @@
+node_modules
+package.json
+coverage
+.nyc_output
+jsHelper.js
```
Comment (mattdean-digicatapult) [inline .eslintignore]: Remove

File: README.md [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+## Environment Variables
+
+`vitalam-service-template` is configured primarily using environment variables as follows:
+
+| variable                      | required | default | description                                                                                     |
+| :---------------------------- | :------: | :-----: | :---------------------------------------------------------------------------------------------- |
+| SERVICE_TYPE                  |    N     | `info`  | Logging level. Valid values are [`trace`, `debug`, `info`, `warn`, `error`, `fatal`]            |
+| PORT                          |    N     | `3001`  | The port for the API to listen on                                                               |
+| API_HOST                      |    Y     |    -    | The hostname of the `vitalam-node` the API should connect to                                    |
```
Comment (mattdean-digicatapult) [inline README.md]: Remove

File: helm/vitalam-service-template/templates/deployment.yaml [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+            - name: PORT
+              valueFrom:
+                configMapKeyRef:
+                  name: {{ include "vitalam-service-template.fullname" . }}-config
+                  key: port
+            - name: API_HOST
+              valueFrom:
+                configMapKeyRef:
+                  name: {{ include "vitalam-service-template.fullname" . }}-config
+                  key: apiHost
```
Comment (mattdean-digicatapult) [inline helm/vitalam-service-template/templates/deployment.yaml]: Remove

File: helm/vitalam-service-template/values.yaml [hunk 21]
```diff
@@ -0,0 +1,8 @@
+config:
+  port: 80
+  apiHost: 'localhost'
```
Comment (mattdean-digicatapult) [inline helm/vitalam-service-template/values.yaml]: Remove

File: helm/vitalam-service-template/templates/configmap.yaml [hunk 22]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "vitalam-service-template.fullname" . }}-config
+  labels:
+    {{- include "vitalam-service-template.labels" . | nindent 4 }}
+data:
+  port: {{ .Values.config.port | quote }}
+  apiHost: {{ .Values.config.apiHost | quote }}
```
Comment (mattdean-digicatapult) [inline helm/vitalam-service-template/templates/configmap.yaml]: Remove

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
Just a few bits that need removing from previous comments

**Review by mattdean-digicatapult — APPROVED** [review]

**Review by jonmattgray — APPROVED** [review]

**Review by dblane-digicatapult — APPROVED** [review]

**Pull request merged**